astropy/units/format/utils.py

```diff
@@ -4,27 +4,41 @@
 Utilities shared by the different formats.
 """
 
+from __future__ import annotations
+
 import warnings
+from typing import TYPE_CHECKING
 
 from astropy.units.utils import maybe_simple_fraction
 from astropy.utils.misc import did_you_mean
 
+if TYPE_CHECKING:
+    from collections.abc import Callable, Container, Mapping, Sequence
+    from numbers import Real
+    from typing import TypeVar
+
+    from astropy.units import UnitBase
+
+    T = TypeVar("T")
+
 
-def get_grouped_by_powers(bases, powers):
+def get_grouped_by_powers(
+    bases: Sequence[T], powers: Sequence[int]
+) -> tuple[list[tuple[T, int]], list[tuple[T, int]]]:
     """
     Groups the powers and bases in the given
     `~astropy.units.CompositeUnit` into positive powers and
     negative powers for easy display on either side of a solidus.
 
     Parameters
     ----------
-    bases : list of `astropy.units.UnitBase` instances
+    bases : `~collections.abc.Sequence` of `astropy.units.UnitBase`
```

Comment:
I think this can just be "sequence of |Unit|" - numpydoc knows what "sequence" is.

Reply:
I decided not to edit this at all. I don't think it's really a problem if the annotation says `Sequence`, but the docstring says `list`. If somebody does want to tweak the docstring they can do it in a follow-up pull request.